packages/contracts/contracts/L1/GasMetering.sol

```diff
@@ -0,0 +1,94 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import { console } from "forge-std/console.sol";
+
+import { SlotPacking128x64x64 } from "../libraries/SlotPacking128x64x64.sol";
+import { SlotPacking128x128 } from "../libraries/SlotPacking128x128.sol";
+
+// TODO: update the spec with nicer variable names and then
+// use the variable names 1:1 in the logic so that its easy
+// to verify
+abstract contract GasMetering {
+    // 128 bits basefee
+    // 64 bites blocknumber
+    // 64 bits boughtGas
+
+    // basefee, blocknumber, boughtGas
+    bytes32 internal storage1;
+
+    // gaslimit, sane gas limit
+    // TODO: need a method for setting these values
+    bytes32 internal storage2;
+
+    uint8 internal constant ELASTICITY_MULTIPLIER = 2;
+    uint8 internal constant BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+
+    function _initialize(
+        uint128 prevBaseFee,
+        uint64 prevNum,
+        uint64 prevBoughtGas,
+        uint128 gasTargetLimit,
+        uint128 gasSanityLimit
+    ) internal {
+        require(storage1 == bytes32(0) && storage2 == bytes32(0));
+
+        storage1 = SlotPacking128x64x64.set(prevBaseFee, prevNum, prevBoughtGas);
+
+        storage2 = SlotPacking128x128.set(gasTargetLimit, gasSanityLimit);
```

Comment:
The `gasTargetLimit` and `gasSanityLimit` were technically defined as `uint64`s in the spec, I made them into `uint128`s out of convenience. If the must be `uint64`s then I will update this code